pkgs/top-level/all-packages.nix

```diff
@@ -8653,15 +8653,19 @@ with pkgs;
 
   libgsf = callPackage ../development/libraries/libgsf { };
 
-  # glibc provides libiconv so systems with glibc don't need to build libiconv
-  # separately, but we also provide libiconvReal, which will always be a
-  # standalone libiconv, just in case you want it
-  libiconv = if stdenv ? cross then
-    (if stdenv.cross.libc == "glibc" then libcCross
-      else if stdenv.cross.libc == "libSystem" then darwin.libiconv
-      else libiconvReal)
-    else if stdenv.isGlibc then glibcIconv stdenv.cc.libc
-    else if stdenv.isDarwin then darwin.libiconv
+  # GNU libc provides libiconv so systems with glibc don't need to build
+  # libiconv separately. Additionally, Apple forked/repackaged libiconv so we
+  # use that instead of the vanilla version on that OS.
+  #
+  # We also provide `libiconvReal`, which will always be a standalone libiconv,
+  # just in case you want it regardless of platform.
+  libiconv =
+    if hostPlatform.libc == "glibc"
+      then glibcIconv (if hostPlatform != buildPlatform
+                       then libcCross
+                       else stdenv.cc.libc)
+    else if hostPlatform.isDarwin
+      then darwin.libiconv
```

Comment:
A very similar case to this is ```libintlOrEmpty``` where on glibc gettext/libintl is included but everywhere else it isn't. I really dislike the ```orEmpty``` part, maybe we should give it the same treatment as this? See:

https://github.com/NixOS/nixpkgs/pull/25976/files/fbbab97b1c709f40172c255edc914d03efa6635a..a65c46dc79588ab6c601cd0a20ae59658a8f1104#diff-036410e9211b4336186fc613f7200b12L8680